timeseries-streaming/timeseries-java-applications/Examples/src/main/java/com/google/dataflow/sample/timeseriesflow/examples/fsi/forex/ForexCSVAdaptor.java

```diff
@@ -0,0 +1,190 @@
+
+      @StartBundle
+      public void startBundle() {
+        CsvMapper mapper = new CsvMapper();
+        objectReader =
+            mapper.readerFor(Forex.class).with(Forex.getCsvSchema().withColumnSeparator(','));
+      }
+
+      @ProcessElement
+      public void process(@Element String input, OutputReceiver<TSDataPoint> outputReceiver) {
+        try {
+          MappingIterator<Forex> object = objectReader.readValues(input);
+          if (object.hasNext()) {
+            Forex forex = null;
+            try {
+              forex = object.next();
+
+              if (convertCSVForex.tickers.contains(forex.ticker)) {
+
+                TSKey key = TSKey.newBuilder().setMajorKey(forex.ticker).build();
+
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS");
+
+                LocalDateTime ldt = LocalDateTime.parse(forex.timestamp, formatter);
+                // Setting UTC using Reykjavik time zone
+                ZonedDateTime zdt = ldt.atZone(ZoneId.of("Atlantic/Reykjavik"));
+
+                long millis = zdt.toInstant().toEpochMilli();
+
+                // OUTPUT Ask Price
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("ASK"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.ask))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Bid Price
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("BID"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.bid))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Ask volume
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("ASK_VOLUME"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.ask_volume))
+                        .build(), Instant.ofEpochMilli(millis));
+
+                // OUTPUT Bid volume
+                outputReceiver.outputWithTimestamp(
+                    TSDataPoint.newBuilder()
+                        .setKey(key.toBuilder().setMinorKeyString("BID_VOLUME"))
+                        .setTimestamp(Timestamps.fromMillis(millis))
+                        .setData(CommonUtils.createNumData(forex.bid_volume))
+                        .build(), Instant.ofEpochMilli(millis));
+              }
+            } catch (Exception ex) {
+              System.out.println(String.format("Unable to parse record: %s", ex));
```

Comment:
Consider adding deadletter support. 
Also use LOGGER for messages when inside the transform.

Reply:
Done, added basic deadletter support.